### ISSUE TITLE
re-evaluate pnetcdf for 1-d fields with pio2

### ISSUE THREAD

**Issue by billsacks** [ESCOMP/CTSM#33]
**Bill Sacks < sacks@ucar.edu > - 2015-06-16 10:32:47 -0600**
**Bugzilla Id:** [2185](http://bugz.cgd.ucar.edu/show_bug.cgi?id=2185)
**Bugzilla CC:**  andre@ucar.edu,  erik@ucar.edu,  jedwards@ucar.edu,  mvertens@ucar.edu,  rfisher@ucar.edu, 

Once we start using PIO2 in CLM, we should re-evaluate bug 1730.

In particular, we should try backing out the workaround that was put in place in clm4_5_1_r091, which involved avoiding pnetcdf for 1-d (dov2xy = .false.) fields. Then we should see if the problem documented in bug 1730 still exists.

If it seems to be resolved with pio2, we should back out the workaround.

**Comment by billsacks** [comment]
@jedwards4b found that this workaround was causing very slow performance with PIO2, at least for some tests. Here are my comments (copied from an email):

> (1) Have you investigated whether this might be a symptom of a broader performance issue with pio2 with netcdf (rather than pnetcdf)? For example, if you change one of these tests to use netcdf across the board, is performance significantly worse with pio2 than with pio1? If so, should that underlying issue be addressed? Or does this large performance hit only seem to occur in this unusual case where we're using pnetcdf for some files and netcdf for others?
> 
> (2) I looked a bit at the timings of your recent tests in your scratch directory. It looks like a large amount of the total i/o time is still being spent in <l:PIO:closefile>. Is that expected? It does look like the timings are now similar to those from a recent test suite run I did with pio1, so that's good. An exception is </glade/scratch/jedwards/ERP_P180x2_D_Ld5.f19_g17.I2000Clm50Sp.cheyenne_intel.clm-default.20201015_095341_jb45cc/timing> vs. </glade/scratch/sacks/tests_1005-205751ch/ERP_P180x2_D_Ld5.f19_g17.I2000Clm50Sp.cheyenne_intel.clm-default.GC.1005-205751ch_int/timing>, but maybe that was machine variability.
> 
> (3) I don't think we can remove this workaround on CTSM master until we have (a) switched to using pio2 in all configurations (or we put in a conditional so that the workaround is applied when using pio1 but not when using pio2), and (b) run some testing to verify that this problem of needing netcdf for vector history files no longer exists with pio2. I read back through the original bug report here <http://bugs.cgd.ucar.edu/show_bug.cgi?id=1730> (you may need to be connected to the UCAR VPN to access that), and it reminded me that this problem only showed up in certain circumstances that may not be uncovered by the test suite. So we'd need to run some supplemental testing to verify that the problem is resolved.
> 
> So, assuming that we're satisfied with the answers to (1) & (2), I'd propose that you push a branch to your CTSM fork that only has the removal of this workaround (not part of the existing PR for ctsm). This branch should be based off of the CTSM tag that was in the last CESM beta tag. Then I can run a bit of manual testing to see if this old problem seems to be resolved. Then I can make a branch tag from this and we can point to that branch tag in the testing with pio2. Once everything seems good with that, I can help bring this to CTSM master.

**Comment by billsacks** [comment]
I have looked into this, and can no longer reproduce the problems that were originally reported on yellowstone with vector history files and pnetcdf – with either pio1 or pio2. So maybe this problem was specific to yellowstone (and doesn't appear on cheyenne), or maybe the underlying problem was solved by changes in pio (or CTSM) in the last 5 years or so.

I also found that, for all of the 5 test runs I did, the runtime was significantly faster with pio2 than with pio1. (I didn't check the runtime in the original code base with the workaround still in place – i.e., writing vector history files using netcdf rather than pnetcdf in pio1.)

----

In more detail: I ran most / all of the test cases laid out in the original bug report (1730 in bugzilla). I ran cases for longer than the originally-reported problems. I ran all of these cases from two code bases: (a) Something close to ctsm5.1.dev007, but with the `avoid_pnetcdf` workaround removed (cherry-picking commits 0071378b9 and 674254bf9); (b) a branch that uses pio2 by default and also includes some other changes, including the removal of this workaround and some other changes to `ncdio_pio` (run from commit d44929a7f).

(1) Tests similar to the original reports:

These tests:
- `SMS_Lm1.f09_g17.I2000Clm50BgcCrop.cheyenne_intel`
- `SMS_Lm1_D.f09_g17.I2000Clm50BgcCrop.cheyenne_intel`

with the following in `user_nl_clm`:

```
hist_nhtfrq = -24
hist_dov2xy = .false.
```

(2) Tests similar to what Stefan reported as a reproducer:

`SMS_Lm1_D.f09_g17.I2000Clm50BgcCrop.cheyenne_intel`

with the following in `user_nl_clm`:

```
hist_nhtfrq     = -24,-24
hist_dov2xy     = .true.,.false.
hist_fincl1     = 'TG'
hist_fincl2     = 'TG'
```

And another that's the same but also with:

```
hist_mfilt = 1,1
```

(3) A test like the one in comment 17 in bugzilla:

`ERP_D_Ld5.f09_g17.I2000Clm50Vic.cheyenne_intel.clm-vrtlay`

where I change the `hist` pieces of `user_nl_clm` to:

```
 hist_dov2xy     = .false.
 hist_ndens      = 1
 hist_nhtfrq     =-24
 hist_mfilt      = 1
```
